native/native.tests/tests/org/jetbrains/kotlin/konan/test/blackbox/ProgramNameTest.kt

```diff
@@ -0,0 +1,66 @@
+package org.jetbrains.kotlin.konan.test.blackbox
+
+import org.jetbrains.kotlin.konan.test.blackbox.support.*
+import org.jetbrains.kotlin.konan.test.blackbox.support.compilation.TestCompilationResult.Companion.assertSuccess
+import org.jetbrains.kotlin.konan.test.blackbox.support.util.ClangDistribution
+import org.jetbrains.kotlin.konan.test.blackbox.support.util.compileWithClang
+import org.jetbrains.kotlin.native.executors.runProcess
+import org.junit.jupiter.api.Tag
+import org.junit.jupiter.api.Test
+import java.io.File
+import kotlin.test.assertEquals
+import kotlin.time.Duration.Companion.seconds
+
+@Tag("program-name")
+class ProgramNameTest : AbstractNativeSimpleTest() {
+
+    @Test
+    fun programNameTest() {
+        // 1. Compile main.c to main.cexe
+
+        val cExecutable = buildDir.resolve("main.cexe")
+        compileWithClang(
+            clangDistribution = ClangDistribution.Llvm,
+            sourceFiles = listOf(sourceDir.resolve("main.c")),
+            outputFile = cExecutable,
+            additionalClangFlags = listOf("-Wall", "-Werror"),
+        ).assertSuccess()
+
+        // 2. Compile kotlinPrintEntryPoint.kt to kotlinPrintEntryPoint.kexe
+
+        val kotlinCompilation = compileToExecutableInOneStage(
+            generateTestCaseWithSingleFile(
+                sourceFile = sourceDir.resolve("kotlinPrintEntryPoint.kt"),
+                testKind = TestKind.STANDALONE_NO_TR,
+                extras = TestCase.NoTestRunnerExtras("main")
+            )
+        ).assertSuccess()
+
+        // 3. run main.cexe (with different parameters) to call kotlin executable
+
+        fun validate(expected: String, vararg args: String) {
+            val binaryName = kotlinCompilation.resultingArtifact.executableFile.path
+            val result = runProcess(cExecutable.absolutePath, binaryName, *args) {
```

Comment:
This fails if the test target is different from the host. You can reproduce this by running this test with `-Pkotlin.internal.native.test.target=<target>` Gradle property.

Please use a proper executor through https://github.com/JetBrains/kotlin/blob/8c6966765b1f34dda1a9c3a5678f12f99045a119/native/native.tests/tests/org/jetbrains/kotlin/konan/test/blackbox/support/settings/SettingsExecutor.kt#L18

Reply:
Thank you for this insight. I changed it accordingly.

Reply:
The test fails on linuxArm64 target with:
```text
java.lang.AssertionError: Expected <calling exec...
programName: app
args:>, actual <calling exec...
exec failed>.
  at kotlin.test.DefaultAsserter.fail(DefaultAsserter.kt:16)
  at kotlin.test.Asserter$DefaultImpls.assertTrue(Assertions.kt:694)
  at kotlin.test.DefaultAsserter.assertTrue(DefaultAsserter.kt:11)
  at kotlin.test.Asserter$DefaultImpls.assertEquals(Assertions.kt:713)
  at kotlin.test.DefaultAsserter.assertEquals(DefaultAsserter.kt:11)
  at kotlin.test.AssertionsKt__AssertionsKt.assertEquals(Assertions.kt:63)
  at kotlin.test.AssertionsKt.assertEquals(Unknown Source)
  at kotlin.test.AssertionsKt__AssertionsKt.assertEquals$default(Assertions.kt:62)
  at kotlin.test.AssertionsKt.assertEquals$default(Unknown Source)
  at org.jetbrains.kotlin.konan.test.blackbox.ProgramNameTest.programNameTest$validate(ProgramNameTest.kt:49)
  at org.jetbrains.kotlin.konan.test.blackbox.ProgramNameTest.programNameTest(ProgramNameTest.kt:54)
...
```

To reproduce, you can run the test with `-Pkotlin.internal.native.test.target=linuxArm64` Gradle property on a Linux/x86_64 machine. I don't have access to such a machine at the moment. Could you please check the test, if possible?

Our test infrastructure runs tests on `linuxArm64` through qemu, maybe the problem is somehow related to that.

Reply:
I could reproduce the problem by running the test via `./gradlew :native:native.tests:test --tests "org.jetbrains.kotlin.konan.test.blackbox.ProgramNameTest" -Pkotlin.internal.native.test.target=linux_arm64` on an x86 linux machine. However you are right, the current test infrastructure with qemu user mode emulation seems to not handle the `exec` syscall in the way we need it. I also tried upgrading qemu to version 9.0.0, but this also didn't resolve the issue. If I understand the following qemu-comment correctly, qemu can't `exec` into a new process which runs again on qemu, but always does a syscall to the host system. This also explains why exec fails with error code 8 (Exec format error), as qemu would instruct the x86 host system to run the arm64 binary: 
> at the point of execve the process leaves QEMU's control
([qemu syscall.c](https://github.com/qemu/qemu/blob/59272469bd1365564fe0bb2c10d8c1d25acd51a6/linux-user/syscall.c#L8520))

I see now some ways to approach the problem:
A) Disable the testcase in case of `testRunSettings.executor is EmulatorExecutable`. Note that [this](https://github.com/JetBrains/kotlin/pull/5281/files#diff-25af118781843da3873165655f4aabc9af76c6080304fce9039a2c707ea0ec62R141) testcase already tests the handling of the "default" program name from the `exec` syscall the JVM does for us. So we already know that reading a value from `argv[0]` works. Then this PR could probably be merged when no new points come up.
B) Create a FullHostSystemEmulatorExecutable, which uses qemu system emulation instead of qemu user space emulation. I'm not sure if this wouldn't be way out of scope for this PR.

Do you have any other ideas? If not, are you fine with approach A?

--
I also managed to run the EmulatorExecutable with the following command, which just does _qemu -> main.cexe -> qemu -> app.kexe_: `/.../qemu-aarch64 -L /.../sysroot /.../main.cexe /.../qemu-aarch64 qemu-aarch64 -L /.../sysroot /.../app.kexe` (minimal refactoring needed in EmulatorExecutable.kt). However, also in this case we can not control the `exec` syscall that actually starts `app.kexe`, as this is part of the qemu user mode emulation infrastrucutre. There qemu sets then manually `argv[0]=app` when launching app.kexe.

Reply:
> Do you have any other ideas? If not, are you fine with approach A?

No ideas. I absolutely fine with it. Moreover, it matches our approach to testing linuxArm64: https://kotlinlang.org/docs/native-target-support.html#tier-2
> The target is regularly tested on CI to be able to compile, but may not be automatically tested to be able to run.